flink-runtime/src/main/java/org/apache/flink/runtime/rest/RestClient.java

```diff
@@ -131,6 +133,11 @@ public class RestClient implements AutoCloseableAsync {
 
     private final String urlPrefix;
 
+    // Used to track unresolved request futures in case they need to be resolved when the client is
+    // closed
+    private final Map<CompletableFuture<Channel>, Void> responseChannelFutures =
+            Collections.synchronizedMap(new IdentityHashMap<>());
```

Comment:
I had gone with `IdentityHashMap` since we want to maintain references to the actual future objects, however since `CompletableFuture`'s `equals` and `hashCode` just delegate to `Object`, the behavior is the same. I'll make the change.